File: cpp/medium/GetImportance.cpp
#include <vector>
using namespace std;
class Employee {
public:
  int id;
  int importance;
  vector<int> subordinates;
};

int getImportance(vector<Employee *> employees, int id) {
  Employee *point = nullptr;
  for (auto e : employees) {
    if (e->id == id) {
      point = e;
      break;
    }
  }
  if (point == nullptr)
    return 0;
  int rs = point->importance;
  vector<int> sub = point->subordinates;
  int sub_index = 0;
  while (sub_index < sub.size()) {
    for (auto e : employees) {
      if (e->id == sub[sub_index]) {
        if (!e->subordinates.empty())
          sub.insert(sub.end(), e->subordinates.begin(), e->subordinates.end());
        rs += e->importance;
        break;
      }
    }
    sub_index++;
  }
  return rs;
}

File: go/middle/getImportance.go
package main

type Employee struct {
	Id           int
	Importance   int
	Subordinates []int
}

func getImportance(employees []*Employee, id int) int {
	idToIndex := make(map[int]int)
	for i := range employees {
		idToIndex[employees[i].Id] = i
	}
	visited := make([]bool, len(employees))
	rs := 0
	var dfs func(cur int)
	dfs = func(cur int) {
		if visited[cur] {
			return
		}
		visited[cur] = true
		rs += employees[cur].Importance
		for _, v := range employees[cur].Subordinates {
			dfs(idToIndex[v])
		}
	}
	dfs(idToIndex[id])
	return rs
}
